hclsyntax/expression_test.go

```diff
@@ -1957,6 +1957,34 @@ EOT
 			cty.ListValEmpty(cty.String), // deduced through refinements
 			0,
 		},
+		{
+			`unknown ? ar : br`,
+			&hcl.EvalContext{
+				Variables: map[string]cty.Value{
+					"unknown": cty.UnknownVal(cty.Bool),
+					"ar": cty.UnknownVal(cty.Set(cty.String)).Refine().
+						CollectionLengthLowerBound(1).CollectionLengthUpperBound(2).NewValue(),
+					"br": cty.UnknownVal(cty.Set(cty.String)).Refine().
+						CollectionLengthLowerBound(3).CollectionLengthUpperBound(4).NewValue(),
+				},
+			},
+			cty.UnknownVal(cty.Set(cty.String)).Refine().NotNull().CollectionLengthLowerBound(1).CollectionLengthUpperBound(4).NewValue(), // deduced through refinements
+			0,
+		},
+		{
+			`unknown ? amr : bmr`,
+			&hcl.EvalContext{
+				Variables: map[string]cty.Value{
+					"unknown": cty.UnknownVal(cty.Bool),
+					"amr": cty.UnknownVal(cty.Set(cty.String)).Mark("test").Refine().
+						CollectionLengthLowerBound(1).CollectionLengthUpperBound(2).NewValue(),
+					"bmr": cty.UnknownVal(cty.Set(cty.String)).Mark("test").Refine().
+						CollectionLengthLowerBound(3).CollectionLengthUpperBound(4).NewValue(),
+				},
+			},
+			cty.UnknownVal(cty.Set(cty.String)).Refine().NotNull().CollectionLengthLowerBound(1).CollectionLengthUpperBound(4).NewValue().Mark("test"), // deduced through refinements
```

Comment:
Post-approval little followup:

Where did the `NotNull` refinement come from here? It doesn't seem like either of the input values are refined as being not-null.